### PR TITLE
[PKG-2813] Rebuild pre-commit 3.2.1 to fix win64

### PR DESCRIPTION
**Pull request by skupr-anaconda**
Changelog: https://github.com/pre-commit/pre-commit/blob/v3.2.1/CHANGELOG.md
License: https://github.com/pre-commit/pre-commit/blob/v3.2.1/LICENSE
Requirements: https://github.com/pre-commit/pre-commit/blob/v3.2.1/setup.cfg

Actions:
1. Increase the build number to `1`
2. Add `--no-build-isolation` to build `scripts`
3. Rename package to `pre-commit` because `conda-build` incorrectly creates `entry_ponits` on `win64`. I guess it's a **conda-build** bug. I left a comment on why not rename it again the next time
4. Move tests from the output `pre-commit` to the top level of the recipe
5. Move test commands from test scripts to the `test/commands` section

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -23,8 +22,8 @@ requirements:
 
 outputs:
   - name: pre-commit
-    script: build_base.bat  # [win]
-    script: build_base.sh  # [not win]
+    build:
+      script: python setup.py install --single-version-externally-managed --record=record.txt
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: I'm not sold on this. I'd prefer if we find the real problem then doing this.

Comment (skupr-anaconda) [reply recipe/meta.yaml]: My suggestion is that the issue is caused by `pip` or `virtualenv` (or by incompatible versions) but I haven't any proves yet

Comment (skupr-anaconda) [reply recipe/meta.yaml]: If I install `pre-commit` **v3.4.0** from conda-forge by `conda install conda-forge::pre-commit` then all dependencies are from the `main` channel:
```
  cffi               pkgs/main/win-64::cffi-1.15.1-py39h2bbff1b_3
  cfgv               pkgs/main/win-64::cfgv-3.3.1-py39haa95532_0
  distlib            pkgs/main/win-64::distlib-0.3.6-py39haa95532_0
  filelock           pkgs/main/win-64::filelock-3.9.0-py39haa95532_0
  identify           pkgs/main/win-64::identify-2.5.5-py39haa95532_0
  nodeenv            pkgs/main/win-64::nodeenv-1.7.0-py39haa95532_0
  platformdirs       pkgs/main/win-64::platformdirs-2.5.2-py39haa95532_0
  pre-commit         conda-forge/noarch::pre-commit-3.4.0-pyha770c72_0
  pycparser          pkgs/main/noarch::pycparser-2.21-pyhd3eb1b0_0
  pyyaml             pkgs/main/win-64::pyyaml-6.0-py39h2bbff1b_1
  ukkonen            pkgs/main/win-64::ukkonen-1.0.1-py39h59b6b97_0
  virtualenv         pkgs/main/win-64::virtualenv-20.17.1-py39haa95532_0
  yaml               pkgs/main/win-64::yaml-0.2.5-he774522_0
```
and the `pre-commit --help` command works. But If I install **v3.2.1** from `main` then the command doesn't work. I wonder what is the difference that can cause the issue.

Comment (cbouss) [reply recipe/meta.yaml]: cf pre-commit is a noarch package. It doesn't include file pre-commit.exe, unlike our package.

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: `pre_commit` is noarch but `pre-commit` has a per python build.

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml]: Ah wait, by `cf` you mean conda-forge?

Comment (cbouss) [reply recipe/meta.yaml]: yes, if you extract both the difference is there

Comment (skupr-anaconda) [reply recipe/meta.yaml]: If I make a noarch package then the error will be the same

**Comment by skupr-anaconda** [comment]
@JeanChristopheMorinPerso @cbouss Finally, I've fixed the issue on `win-64` by restructuring the recipe. 
My way of thinking was:
1. Build the one-output package without the `outputs` section. It hadn't any errors.
2. Then I renamed the top-level `package/name` from `pre-commit-split` to `pre-commit`. It worked fine but tests skipped for the `pre-commit` output but not for the `pre_commit` one.
3. I moved `tests` from the `pre-commit` output to the top level of the recipe. They passed.

The lesson learned:
- `anaconda-linter` provided incorrect advice for this multi-output recipe, **or** 
- `conda-build` has a bug dealing with multi-output packages.

**Comment by skupr-anaconda** [comment]
Staging artifacts are here https://staging.continuum.io/prefect/fs/pre-commit-feedstock/pr5/435da45.
Test it by:
```
conda create -n test
conda activate test
conda install -c "https://staging.continuum.io/prefect/fs/pre-commit-feedstock/pr5/435da45" --repodata-fn repodata.json pre-commit
pre-commit -V
pre-commit --help
```

**Comment by skupr-anaconda** [comment]
Rebuilding with `cfgv 3.4.0` https://github.com/AnacondaRecipes/cfgv-feedstock/pull/2

**Comment by cbouss** [comment]
> @JeanChristopheMorinPerso @cbouss Finally, I've fixed the issue on `win-64` by restructuring the recipe. My way of thinking was:
> 
> 1. Build the one-output package without the `outputs` section. It hadn't any errors.
> 2. Then I renamed the top-level `package/name` from `pre-commit-split` to `pre-commit`. It worked fine but tests skipped for the `pre-commit` output but not for the `pre_commit` one.
> 3. I moved `tests` from the `pre-commit` output to the top level of the recipe. They passed.
> 
> The lesson learned:
> 
> * `anaconda-linter` provided incorrect advice for this multi-output recipe, **or**
> * `conda-build` has a bug dealing with multi-output packages.

Oh I see what happened now. 
In that case anaconda-linter provides the advice that is best in most cases, and conda-build is just complex.
Had the entry_points been defined on the output level instead of the global level, this would have worked.

**Comment by skupr-anaconda** [comment]
> Oh I see what happened now. In that case anaconda-linter provides the advice that is best in most cases, and conda-build is just complex. Had the entry_points been defined on the output level instead of the global level, this would have worked.

I tried to define entry_points on the output level but it's now working at all

**Comment by JeanChristopheMorinPerso** [comment]
Looking at the diff, the most notable change is the test section which was moved to the top level of the build recipe. How does that fixes the issue that you were initially trying to solve (ie pre-commit.exe failing to run)?

**Comment by skupr-anaconda** [comment]
> Looking at the diff, the most notable change is the test section which was moved to the top level of the build recipe. How does that fixes the issue that you were initially trying to solve (ie pre-commit.exe failing to run)?

My first attempt was to rename the package from `pre-commit-split` to `pre-commit`. I think `conda-build` badly interprets the name of the package and the output one at the same time. It builds fine if the package name **equals** the name of the output. If it's **NOT** equal then entry_points on win will be created incorrectly. That's why I think it's conda-build bug but not the upstream one

**Comment by JeanChristopheMorinPerso** [comment]
Thanks for clarifying this. We should create an issue on the conda-build repo if we think there is a bug.

**Comment by cbouss** [comment]
By moving entry_points to the output, it does correct the issue. It results in two files:
Scripts\pre-commit.exe and Scripts\pre-commit-script.py

**Comment by skupr-anaconda** [comment]
> By moving entry_points to the output, it does correct the issue. It results in two files: Scripts\pre-commit.exe and Scripts\pre-commit-script.py

Let me check again...

**Comment by skupr-anaconda** [comment]
> By moving entry_points to the output, it does correct the issue. It results in two files: Scripts\pre-commit.exe and Scripts\pre-commit-script.py

Thanks, Charles. You're right. I **wrongfully** placed entry_points to the output:
```
outputs:
  - name: pre-commit
    script: build_base.bat  # [win]
    script: build_base.sh  # [not win]
    entry_points:
      - pre-commit = pre_commit.main:main
```

The correct way is:
```
outputs:
  - name: pre-commit
    script: build_base.bat  # [win]
    script: build_base.sh  # [not win]
    build:
      entry_points:
        - pre-commit = pre_commit.main:main
```

**Review by cbouss — APPROVED** [review]

**Review by JeanChristopheMorinPerso — APPROVED** [review]

**Pull request merged**